Tools/autotest: add quadplane output_motor_mask check

kd0aij opened the pull request:
This tests whether output_motor_mask works when booting into MANUAL and FBWA modes.
Changed default mode to from FBWA to MANUAL to avoid possibility of transition code becoming active at boot.
Test does fail if I revert commit 35928a8: AP_Motors: fix regression for tiltrotors

peterbarker commented on Tools/autotest/quadplane.py:
```diff
@@ -103,6 +103,40 @@ def get_disarm_delay(self):
 
+    def test_motor_mask(self):
+        """Check operation of output_motor_mask"""
+        if not(int(self.get_parameter('Q_TILT_MASK')) & 1):
```
Does setting this really require a reboot?  The parameter is not marked as such.

kd0aij replied:
being paranoid here: wanted to guarantee that the copter state hasn't been touched

peterbarker replied:
It would be nice to have a fresh vehicle before each test.  However, it's kind of too expensive time-wise to do so.

We do do some work to try to give each test a vehicle in set state - disarmed, on ground, with default parameters.

peterbarker commented on Tools/autotest/quadplane.py:
```diff
@@ -103,6 +103,40 @@ def get_disarm_delay(self):
+    def test_motor_mask(self):
+        """Check operation of output_motor_mask"""
+        if not(int(self.get_parameter('Q_TILT_MASK')) & 1):
+#        or (int(self.get_parameter('Q_TAILSIT_MOTMX')) & 1)
```
Dead code

kd0aij replied:
https://github.com/ArduPilot/ardupilot/pull/10401 would add a new parameter: Q_TAILSIT_MOTMX
that would also result in invoking output_motor_mask

peterbarker replied:
Good-o.  We ordinarily don't like dead code, but *shrug*.

peterbarker commented on Tools/autotest/quadplane.py:
```diff
@@ -103,6 +103,40 @@ def get_disarm_delay(self):
+        assert(33 == self.get_parameter('SERVO5_FUNCTION'))
+
+        self.progress("Testing MANUAL mode")
+        self.mavproxy.send('mode MANUAL\n')
```
I prefer `change_mode` now-adays.

kd0aij replied:
didn't notice change_mode, will look at it

peterbarker commented on Tools/autotest/quadplane.py:
```diff
@@ -103,6 +103,40 @@ def get_disarm_delay(self):
+        self.mavproxy.send('mode MANUAL\n')
+        self.wait_mode('MANUAL')
+        self.arm_vehicle()
+        self.progress("Raising throttle to 1800")
```
Ah, a number to get out of date :-)

kd0aij replied:
Is there a "proper" way to make motors spin?

peterbarker replied:
Nah, just waiting for the numbers to diverge :-)

peterbarker commented on Tools/autotest/quadplane.py:
```diff
@@ -103,6 +103,40 @@ def get_disarm_delay(self):
+        self.progress("Waiting for Motor1 to start")
+        self.wait_servo_channel_exceed_value(5, 1100)
+
+        self.reboot_sitl()
```
Would be nice to avoid the reboot here.  Drop the servo, wait for it to reduce, then continue?

kd0aij replied:
"drop" a servo?

peterbarker replied:
Sorry, make the servo output channel return to its normal value :-)

peterbarker commented on Tools/autotest/quadplane.py:
```diff
@@ -103,6 +103,40 @@ def get_disarm_delay(self):
+        self.progress("Raising throttle to 1800")
+        self.set_rc(3, 1800)
+        self.progress("Waiting for Motor1 to start")
+        self.wait_servo_channel_exceed_value(5, 1100)
```
Leaving the vehicle armed with motor spinning at the end of the test isn't good :-)

Please `self.zero_throttle()` and `self.disarm_motors()`

kd0aij replied:
A test should leave the vehicle in some specific state?

peterbarker replied:
Yes, disarmed on the ground.

If it's not then we will reboot the vehicle, which slows the test suite own.

peterbarker commented on Tools/autotest/common.py:
```diff
@@ -1323,6 +1323,28 @@ def wait_servo_channel_value(self, channel, value, timeout=2):
             if m_value == value:
                 return
 
+    def wait_servo_channel_exceed_value(self, channel, value, timeout=2):
```
This is fine, but if we merge it as-is I might come back later and add an optional `operator=operator.eq` keyword parameter to `wait_servo_channel_value`, based on https://stackoverflow.com/questions/18591778/how-to-pass-an-operator-to-a-python-function

The call to `self.wait_servo_channel_exceed_value(5, 1100)` would become `self.wait_servo_channel_value(5, 1100, operator=operator.gt)`

kd0aij replied:
I'd also prefer a more elegant approach here

peterbarker commented:
On Mon, 11 Feb 2019, Mark Whitehorn wrote:

> +        self.progress("Raising throttle to 1800")
> 
> Is there a "proper" way to make motors spin?

Nah.  But I'll get back to you when the code diverges from the comment ;-)

kd0aij commented:
@peterbarker I think this addresses all of your review comments (finally understood your problem with that comment :) ) and verified that the reboots weren't necessary to catch the regression.  
Not sure if the final state is acceptable yet  though; I'm assuming that disarm returns the motor outputs to miniimum or zero.

peterbarker commented on Tools/autotest/common.py:
```diff
@@ -1301,26 +1302,27 @@ def wait_distance(self, distance, accuracy=5, timeout=30):
-    def wait_servo_channel_value(self, channel, value, timeout=2):
-        """wait for channel to hit value"""
+    def wait_servo_channel_value(self, channel, value, timeout=2, comparator=operator.eq):
+        """wait for channel to exceed value"""
```
Comment is incorrect, esp. with the default operator.

Also the case with other comments in here.

peterbarker commented on Tools/autotest/common.py:
```diff
@@ -1301,26 +1302,27 @@ def wait_distance(self, distance, accuracy=5, timeout=30):
+    def wait_servo_channel_value(self, channel, value, timeout=2, comparator=operator.eq):
+        """wait for channel to exceed value"""
         channel_field = "servo%u_raw" % channel
+        opstring = ("%s"%comparator)[-3:-1]
```
I'm *sure* flake8 is going to complain about this line because of the lack of whitespace around `%` :-)

`flake8 common.py` will lint this file for you.  I've no doubt unrelated complaints will pop up at the moment on common.py

peterbarker commented on Tools/autotest/common.py:
```diff
@@ -1436,7 +1438,7 @@ def wait_mode(self, mode, timeout=60):
     def wait_ready_to_arm(self, timeout=None, require_absolute=True):
         # wait for EKF checks to pass
-        self.progress("Waiting reading for arm")
+        self.progress("Waiting ready for arm")
```
`s/for/to/g`

peterbarker commented:
On Tue, 12 Feb 2019, Mark Whitehorn wrote:

> Not sure if the final state is acceptable yet though; I'm assuming that
> disarm returns the motor outputs to miniimum or zero.

It will.  We do also reset the RC inputs between tests.

peterbarker commented:
Merged, thanks!
